### ISSUE TITLE
[Potential NAN bug] Loss may become NAN during training

### ISSUE THREAD

**Issue by Justobe** [leosky21/python-tensorflow#1]
Hello~

Thank you very much for sharing the code!

I try to use my own data set ( with the same shape as mnist) in  code. After some iterations, it is found that the training loss becomes NAN. After carefully checking the code, I found that the following code may trigger NAN in loss:

In `python-tensorflow/tensorflow_MINIST/__init_MNIST01.py`

```python
cross_entropy = tf.reduce_mean(- tf.reduce_sum(y * tf.log(prediction), reduction_indices=[1]))
```

If  prediction contains 0 (output of softmax ), the result of `tf.log(prediction)` is `inf` because `log(0)` is illegal . And this may cause the result of loss to become NAN.

It could be fixed by making the following changes:

```python
cross_entropy = tf.reduce_mean(- tf.reduce_sum(y * tf.log(prediction + 1e-10), reduction_indices=[1]))
```

or

```python
cross_entropy = tf.reduce_mean(- tf.reduce_sum(y * tf.log(tf.clip_by_value(prediction,1e-10,1.0)), reduction_indices=[1]))
```

Hope to hear from you ~

Thanks in advance!  : )
